akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala

```diff
@@ -775,7 +775,11 @@ private final case class SavedIslandData(
 
   override def takePublisher(slot: Int, publisher: Publisher[Any], publisherAttributes: Attributes): Unit = {
     val connection = conn(slot)
-    val bufferSize = publisherAttributes.mandatoryAttribute[InputBuffer].max
+    // if the publisher is a source of elements, use the buffer size attributes from itself (#30076)
+    // for all other cases, pick up the buffer size from the destination logic
+    val bufferSize =
+      (if (connection.inOwner.outCount == 0) publisherAttributes
```

Comment:
Wouldn't that still wrong if we have something like 

```scala
Source(1,2,3).addAttributes(Attributes.inputBuffer(1, 1)).async
  .map(_ * 12).addAttributes(Attributes.inputBuffer(2, 2))
```

where it would use the first attribute instead of the second? If we want to use something like that it would have to be specialized to `PublisherSource` and `SubscriberSource`.